### ISSUE TITLE
systemd-timers skipped timers scheduled between 2AM and 3AM on Sunday DST change (March)

### ISSUE THREAD

**Issue by ntayachi** [systemd/systemd#5595]
### Submission type

  - [ ] Bug report
  - [x] Request for enhancement (RFE)

### systemd version the issue has been seen with

> systemd 231

### Used distribution

> NixOS 16.09

### Description
I am scheduling a task to run every day at 2AM, and on Sunday March 12th with the DST change the clock jumped from 2AM to 3AM, and therefore my task didn't run.
It would be great if systemd timers takes into consideration the DST change. 

When I typed systemctl list-timers, I got this:
```
NEXT                         LEFT          LAST                         PASSED       UNIT                             ACTIVATES
Mon 2017-03-13 02:00:00 EDT  22h left      Sat 2017-03-11 02:00:02 EST  24h ago      my_task.timer                my_task.service
```

**Comment by zdzichu** [comment]
There was no 2AM hour on March 12th, right?

**Comment by ntayachi** [comment]
Yes you're correct there was no 2AM.
But I know that crontab takes into consideration this DST change, and it run immediately the skipped jobs. (please see https://linux.die.net/man/8/cron)

> If time has moved forward, those jobs that would have run in the interval that has been skipped will be run immediately. Conversely, if time has moved backward, care is taken to avoid running jobs twice.

Is it feasible to implement something like this for systemd timers?

**Comment by func0der** [comment]
Can't you put a timezone in the onCalenader configuration?

For me this behavior seems correct, because as @zdzichu said: There was no 2pm on March 12th

**Comment by ntayachi** [comment]
Hi @func0der,

Thank you very much for your answer!
I am not aware that we can add timezone to OnCalendar. I will give it a try.
However, I have another question: 
In my case, I need to run my_task at 2 am every day regardless of DST, which timezone should I put in OnCalendar: EST?

Thanks,
Nafaa

**Comment by func0der** [comment]
@nafaatayachi
The time in OnCalendar ist expected to be in the servers timezone.
So if you really want to be 100% sure, use UTC.
I do not think the behaviour would change, if you add your timezone to the onCalendar setting, due to the fact that there still would not be a 2pm on March 12th.

**Comment by zerkms** [comment]
> Is it feasible to implement something like this for systemd timers?

It's not.

Let's say you scheduled to run something at 11:00am every 31st day of the month. Do you it to ever run in February or April?

And `February 29` - should it be run in a non-leap year?

> In my case, I need to run my_task at 2 am every day regardless of DST, which timezone should I put in OnCalendar: EST?

2am sometimes happens twice a day (and sometimes does not happen at all). Your task explanation implies it should run twice. Otherwise - it should be formalised more explicitly to somehow exclude it. But how about `Etc/GMT-5`? Or any other timezone that does not have summer time. But, in that case it would still be run in different local hours.

**Comment by nafaatayachi** [comment]
Hello,

Sorry for the late reply.
I was able to find a workaround for my issue by simply scheduling my task at 01:55am.
A Friend of mine faced the same issue as well, she had a timer scheduled at 02:15am and she was able to find a workaround for that by creating another timer as follows:

```
my_job_dst_change
[Timer]
OnCalendar=Sun *-03-8..14 03:15
```

It's basically scheduling the second job "my_job_dst_change" on the second Sunday of March i.e. between 8th and 14th of March at 03:15am.
Posting this just in case someone else is facing this issue.

Finally, if handling DST automatically is still not feasible in systemd then I guess this issue can be closed.

Thanks!

**Comment by nafaatayachi** [comment]
Actually it could be simpler, by using one timer (the original one) and adding another OnCalendar
```
my_job
[Timer]
OnCalendar=02:15
OnCalendar=Sun *-03-8..14 03:15
```

**Comment by zerkms** [comment]
> Finally, if handling DST automatically is still not feasible in systemd then I guess this issue can be closed.

It is handled properly. If time shifts overnight then `02:15am` never happens (it does not exist) so the timer is not triggered.

**Comment by nafaatayachi** [comment]
Sorry for the confusion @zerkms.
By: 

> handling DST automatically

I meant:

> NOT skip jobs between 02:00 and 02:59

**Comment by zerkms** [comment]
I'd say it's impossible without making it counter intuitive.

**Comment by zootboy** [comment]
@nafaatayachi Your 1:55 timer might run into the opposite problem on the other end of DST. The actual behavior of exiting DST is for the time to roll from 1:59:59 to 1:00:00, thus any timers in the 1:00:00 to 1:59:59 period may run twice (I haven't verified this, but I would assume it's the case).

**Comment by OhNoMoreGit** [comment]
> The actual behavior of exiting DST is for the time to roll from 1:59:59 to 1:00:00, thus any timers in the 1:00:00 to 1:59:59 period may run twice (I haven't verified this, but I would assume it's the case).

Actually, the current behaviour is that ''no'' timers execute during the "duplicate" hour. See #8647.

**Comment by septatrix** [comment]
> Actually, the current behaviour is that _no_ timers fire during the "duplicate" hour. See #8647.

This I think is problematic in that it means that different ends of daylight saving time are handled differently. I would either expect the behaviour to be like crontab (i.e. fire skipped timers and do not run duplicates) or to not handle summertime at all (i.e. do not run timers when the time skips forward and run them twice when time goes backwards).

**Comment by arodland** [comment]
Yeah, it looks to me like there are two reasonable behaviors and right now systemd does neither one.

"Aware" behavior: for Standard->DST transition, run timers scheduled during the missing hour afterwards; for DST->Standard transition, don't run timers during the duplicate hour. This generally gives the expected behavior for jobs that have a daily-or-longer cadence. For instance, if you have a backup job set to run daily at 1:30AM, this policy will make sure it runs once per day, not 0 or 2 times, and that the time between runs is (availability permitting), 23, 24, or 25 hours — not 1 or 49. Likewise if it was a weekly or a monthly job that happened to land in the window.

"Unaware" behavior: for Standard->DST transition, ignore timers scheduled during the missing hour; for DST->Standard transition, run timers during both instances of the duplicate hour. This generally gives the expected behavior for jobs that have an hourly-or-shorter cadence. For instance if you have a job set to `OnCalendar=*-*-* *:*/15:00` this policy will make sure it runs four times per hour, not 0 (or 8, though that probably wouldn't happen anyway), and that the time between runs is 15 minutes, not 75.

"Aware" behavior gives a surprising behavior (missing runs on DST exit) for short-cadence timers. "Unaware" behavior gives two surprising behaviors (missing runs on DST entrance and duplicate runs on DST exit) for long-cadence timers. Systemd's current policy is a "half-aware" mix of the two behaviors, and mixes their problems ­— it messes up *both* kinds of timers at least some of the time.

I think that being able to choose a behavior explicitly per-timer would be fine.

I think that choosing a behavior automatically per-timer based on the cadence would be possible, and work right probably 98% of the time.

I think that if neither of those is possible, systemd should stick to one of the two behaviors @septatrix and I enumerated, and recommend that users find another tool for timers that need the other behavior — rather than mixing them as it does now.

**Comment by zerkms** [comment]
@arodland "it messes up both kinds of timers" --- why? It was me who originally implemented timers support to the systemd and to me its current behaviour (now I think as a user that uses that feature too) is the most logical: timers run when clocks show exactly the time that was configured, not earlier or later.

It's the same as - if you set up a timer to be run on the `Feb 29th` it would only run once every 4 years*, and you cannot say it's unexpected, it's exactly what I want - I set up timer to run on the "last day of February only on leap years".

**Comment by OhNoMoreGit** [comment]
@zerkms

When people write, say:

```
OnCalendar=*:*:*/15
```

they almost always mean "every 15 minutes", not "every 15 minutes, except for a particular day on the year when nothing happens for an hour".

**Comment by OhNoMoreGit** [comment]
OK, so I might be getting confused here... and I suspect other people are too. Some regions in the US have just gone through a DST→non-DST transition, which might be why this issue is attracting attention right now... but this issue is actually about the "missing" hour in the non-DST→DST transition (in March in the US).

For the DST→non-DST transition, where the hour is "duplicated", I had actually opened a completely separate issue #8647. However that issue got closed as a duplicate of this issue, even though it's for the _other_ kind of transition. As far as I know that bug still isn't fixed.

**Comment by zerkms** [comment]
@OhNoMoreGit let's imagine you're designing a timer, and you have an instruction

```
run a task every February 29th at 10:00am Pacific/Auckland
```

what will be the next date-and-time when you run it?

**Comment by TriMoon** [comment]
> I am scheduling a task to run every day at 2AM, and on Sunday March 12th with the DST change the clock jumped from 2AM to 3AM, and therefore my task didn't run. It would be great if systemd timers takes into consideration the DST change.
> 
> When I typed systemctl list-timers, I got this:
> 
> ```
> NEXT                         LEFT          LAST                         PASSED       UNIT                             ACTIVATES
> Mon 2017-03-13 02:00:00 EDT  22h left      Sat 2017-03-11 02:00:02 EST  24h ago      my_task.timer                my_task.service
> ```

Might be just me but why is the first noted in `EDT` and second in `EST` :thinking: 
Looking at [edt/est](https://24timezones.com/difference/edt/est) seems there is a 1hr diff between the two...
Also, doesn't `systemd` keep times in `UTC` internally, which would mean DST changes of any timezone should not affect the actual internal period?

**Comment by zerkms** [comment]
Btw, I agree that it's _almost always_ not intended behaviour (to either duplicate, or skip), but I think that if it should be "fixed" - then this behaviour should be enabled _explicitly_, eg: `ResolveDSTSwitchGlitches=true` (think of a better name here, but I hope it's clear what it's fot).

Why explicit option: explicit is better than implicit. What if somebody sets up a timer for something to run at `2:30am` simply because some API is free from `2:25am to 2:35am`. And in other times every API call costs $1000. It would be not cool to run it at 3:00am simply because it skipped due to DST. And in this very artificial scenario it would be better to run exactly only at `2:30am` if ever at all, otherwise a user will owe a lot of money to the API provider.

**Comment by OhNoMoreGit** [comment]
@zerkms 

The next February 29, which might be 4 or 8 years away. Nobody is asking for everything that would have happened on February 29 to happen on March 1 instead. People just want reasonably sane behaviour across DST transitions. That's a duplicate or missing hour, not a whole day. In other words, it only affects timers that fire more frequently than once an hour.

**Comment by TriMoon** [comment]
IMHO, it is the responsibility of the admin, that sets up the timer, to account for DST changes.
Not of `systemd` which is just executing what the admin asked for.
You can't blame a CPU to not be intelligent like a human, it won't refuse to erase your OS if you ask it to right?

So yea IMHO, this DST quirk is not something the code should count for, but instead the admin...

**Comment by zerkms** [comment]
@TriMoon 

> Also, doesn't systemd keep times in UTC internally, which would mean DST changes of any timezone should not affect the actual internal period?

systemd-timers is effectively a pattern matching engine: if time matches the requested pattern - it triggers.

**Comment by TriMoon** [comment]
> @TriMoon
> 
> > Also, doesn't systemd keep times in UTC internally, which would mean DST changes of any timezone should not affect the actual internal period?
> 
> systemd-timers is effectively a pattern matching engine: if time matches the requested pattern - it triggers.

Doesn't it parse those patterns at startup and use UTC internally?

**Comment by OhNoMoreGit** [comment]
Just to add one extra complication to all of this... not all Crons work the same. Cronie, for instance, completely skips the missing hour's jobs.

FWIW, I do think that is better than what Vixie Cron does. In my opinion what systemd is doing now for the missing hour is "more correct" than the alternative.

But people used to Vixie Cron will probably want their systemd timers to work similarly, so having a way to override the behaviour would be good.

**Comment by zerkms** [comment]
> Doesn't it parse those patterns at startup and use UTC internally?

it does, but it does not matter how it's stored internally: after pattern is matched against - it can be any unit (be it UTC or CPU ticks).

> Might be just me but why is the first noted in EDT and second in EST

because in 2017 EST->EDT happens in the March 12th.

**Comment by TriMoon** [comment]
> > Doesn't it parse those patterns at startup and use UTC internally?
> 
> it does, but it does not matter how it's stored internally: after pattern is matched against (and pattern matching respect timezones) - it can be any unit (be it UTC or CPU ticks).

You mean it keeps re-evaluating the patterns each CPU tic? :thinking: :exploding_head: 
Why would a DST change affect the used time to trigger the event, when the actual moment to trigger the event is kept unchanged from the DST-change in memory?
_(The RTC's value doesn't change when DST changes, just the interpretation of it)_
> 
> > Might be just me but why is the first noted in EDT and second in EST
> 
> because in 2017 EST->EDT happens in the March 12th.

Sure but why the **TimeZone** change in the display?
Shouldn't they be printed in same timezone at any given moment?
I mean when it's asked while in the `EST` timezone, it should print both times as `EST`, same with `EDT` :thinking: 
Sure the moment they would trigger could be in different DST but that is irrelevant for the output at any given moment.
Or are my brains really fried right now? :rofl:

**Comment by OhNoMoreGit** [comment]
> You mean it keeps re-evaluating the patterns each CPU tic?

No, when scheduling the timer (i.e. when it's first activated, just after it fires, or just after systemd detects that the system clock has jumped) systemd simply determines the next time that would match the pattern. The internal format in which that time is stored isn't particularly important.

**Comment by zerkms** [comment]
> You mean it keeps re-evaluating the patterns each CPU tic?

No, it's calculated once, but the internal unit used to store does not matter.

> Why would a DST change affect the used time to trigger the event, when the actual moment to trigger the event is kept unchanged from the DST-change in memory?

It does not affect it, the next time to trigger already takes into account DST that will happen in the future.

> Sure but why the TimeZone change in the display?

Because March 13th is in EDT.

> I mean when it's asked while in the EST timezone, it should print both times as EST, same with EDT

Nope, it would be confusing: `November 6th 2022, 1:30am EST` and `November 6th 2022, 1:30am EDT` are different points in time. Even if "now" is still `EST` - then `November 6th 2022, 1:30am EDT` is more logical to see not `November 6th 2022, 2:30am EDT`.

**Comment by zerkms** [comment]
> It should indeed increase the UTC versions instead of the locale timezone's versions

it uses the timezone that was specified in the timer pattern.

**Comment by OhNoMoreGit** [comment]
@zerkms 

But do you really expect people to grok the difference between:

```
OnCalendar=*:30:0
```

and:

```
OnCalendar=*:30:0 UTC
```

?

**Comment by zerkms** [comment]
> Why doesn't the * there just mean "every hour"?

It does mean every hour. Every hour when minutes component `= 30`.

**Comment by OhNoMoreGit** [comment]
No, see [my comment above](https://github.com/systemd/systemd/issues/5595#issuecomment-1306829922). systemd skips one of the duplicate hours.

**Comment by TriMoon** [comment]
@zerkms 
> > It should indeed increase the UTC versions instead of the locale timezone's versions
> 
> it uses the timezone that was specified in the timer pattern.

Ofcourse it should use the timezone of the specified pattern as a start point, but then use the internal UTC versions to populate the next ones...
Because from the point of view of a value, it doesn't make sense to account for the interpreted value.
eg. `x^2` should not change graph when `x` is interpreted differently at any value, makes sense right?

**Comment by zerkms** [comment]
> No, see https://github.com/systemd/systemd/issues/5595#issuecomment-1306829922. systemd skips one of the duplicate hours.

Okay, now that is a bug then

```
TZ=Pacific/Auckland systemd-analyze calendar --base-time=2023-04-02 --iterations=4 '*:30:0'
  Original form: *:30:0                      
Normalized form: *-*-* *:30:00               
    Next elapse: Sun 2023-04-02 00:30:00 NZDT
       (in UTC): Sat 2023-04-01 11:30:00 UTC 
       From now: 4 months 22 days left       
       Iter. #2: Sun 2023-04-02 01:30:00 NZDT
       (in UTC): Sat 2023-04-01 12:30:00 UTC 
       From now: 4 months 22 days left       
       Iter. #3: Sun 2023-04-02 02:30:00 NZDT
       (in UTC): Sat 2023-04-01 13:30:00 UTC 
       From now: 4 months 22 days left       
       Iter. #4: Sun 2023-04-02 03:30:00 NZST
       (in UTC): Sat 2023-04-01 15:30:00 UTC 
       From now: 4 months 22 days left  
```

I expect to see `Sun 2023-04-02 02:30:00 NZST` there too. (which is `Sat 2023-04-01 14:30:00 UTC`)

**Comment by OhNoMoreGit** [comment]
> Okay, now that is a bug then

Great! Can you please re-open #8647 where I first reported it then?

_This_ bug isn't the right place. _This_ bug is about the _other_ DST transition (and, as I said in an earlier comment, I think what systemd is doing is "more correct" than anything else on that transition).

**Comment by zerkms** [comment]
I cannot open it though, I only PRed here once 5 years ago :-D Will leave a comment there though.

**Comment by TriMoon** [comment]
@zerkms no idea where they do the loop to generate the next occurrences of the repeated events of a single timer, but that part is where they should look for bugfixing IMHO.
Especially taking into account what i wrote in https://github.com/systemd/systemd/issues/5595#issuecomment-1306846913

> I cannot open it though, I only PRed here once 5 years ago :-D Will leave a comment there though.

You could ping `@poettering` to grab his attention :smile_cat:

**Comment by zerkms** [comment]
> You could ping @poettering to grab his attention

just did

> no idea where they do the loop to generate the next occurrences of the repeated events of a single timer

probably here: https://github.com/zerkms/systemd/blob/25612ecba42cd45e12438c1f63540a930c6de475/src/basic/calendarspec.c#L1340

**Comment by arodland** [comment]
> It was me who originally implemented timezone timers support to the systemd and to me its current behaviour (now I think as a user that uses that feature too) is the most logical: timers run when clocks show exactly the time that was configured, not earlier or later.

If that was true it would be one of the reasonable alternatives I described (the "unaware" one), but that's *not* what currently happens.

If "what you wrote was what you got", a timer with `OnCalendar=minutely` would run every minute... instead, it runs every minute *most* of the time except when it skips 60!

**Comment by mbiebl** [comment]
Related downstream bug report: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1033548
